Parsing the relevant info from XML

XiaoleiZ opened the pull request:
The src/parse_clinvar_xml.py would extract the relevant fields from ClinVar xml file: the variation type (column name: Measureset_type), variation id (column name: Measureset_id), number of alleles reported in the variantion (column name: number_alleles), allele id, gene_symbol, clinicial_significance, review_status etc. 

To keep the records with just one allele in clinvar.tsv, just filter by the column: number_allele

XiaoleiZ committed:
Update parse_clinvar_xml.py

extract the variant-condition interpretation infos from XML

XiaoleiZ committed:
Update parse_clinvar_xml.py

Add usage

XiaoleiZ committed:
Add files via upload

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

bw2 commented on README.md:
```diff
@@ -2,66 +2,24 @@
 
 #### In 1 sentence
```
So we can merge your changes into master, can you restore the original README and instead of documenting the branch, just edit the few places in the original README that are no longer correct after your changes?

bw2 commented on src/add_exac_fields.py:
```diff
@@ -1,116 +0,0 @@
-"""
-Script for generating a new clinvar table with the ExAC fields below added to each clinvar variant that's in ExAC
```
Please restore this file. We can keep using it to generate the clinvar_with_exac.tsv file

bw2 commented on src/clinvar_table_to_vcf.py:
```diff
@@ -1,119 +0,0 @@
-import argparse
-import collections
-import gzip
```
please restore this file

bw2 commented on src/master.py:
```diff
@@ -1,153 +0,0 @@
-"""
-Alternate implementation of master.bash with improved logging, skipping of commands if output files are up-to-date, parallelization, etc.
-
-Run with -h to see all options.
```
please restore this file and modify it to use your new parse script rather than creating a new master_parse_clinvar_xml.py

bw2 commented on src/normalize.py:
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+
```
please remove this file from git. It's retrieved from another repo at the start of the pipeline.

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -1,4 +1,15 @@
-#!/usr/bin/env python
+'''
+Created on 16 Dec 2016
```
please remove lines 1-12 to match how the other files are structured.

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -20,43 +33,100 @@ def replace_semicolons(s, replace_with=":"):
     header = [
-        'chrom', 'pos', 'ref', 'alt', 'mut', 'measureset_id', 'all_submitters', 'all_traits', 'all_pmids',
-        'inheritance_modes', 'age_of_onset', 'prevalence', 'disease_mechanism', 'xrefs'
+        'chrom', 'pos', 'ref', 'alt', 'mut', 'Measureset_type','Measureset_id','number_alleles','Measure_type','alleleid', 'name','gene_symbol',
```
header fields should be all lower-case. 

alleleid would be better as `allele_id`

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -20,43 +33,100 @@ def replace_semicolons(s, replace_with=":"):
-def parse_clinvar_tree(handle,dest=sys.stdout,verbose=True,mode='collapsed'):
+def parse_clinvar_tree(handle,dest=sys.stdout,error=sys.stderr,verbose=True,mode='collapsed'):
     # print a header row
+    #header = [
```
remove commented out code

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -20,43 +33,100 @@ def replace_semicolons(s, replace_with=":"):
+        'molecular_consequence','variant_type','variant_type_change','all_conditions_name'
+        ,'clinical_significance','review_status','all_submitters'
+        , 'all_pmids','inheritance_modes', 'age_of_onset', 'prevalence', 
+        'disease_mechanism', 'xrefs'
```
other columns from the previous clinvar.tsv that should be added here: 

symbol
pathogenic
benign
conflicted
gold_stars
hgvs_c
hgvs_p
origin

bw2 replied:
actually, I see the 'name' column contains both hgvs_c and hgvs_p values 
`'name': 'NM_014855.2(AP5Z1):c.1413_1426delGGACCTGCCCTGCT (p.Leu473Glyfs)',`

it would be good to split these into 2 separate columns like in https://github.com/macarthur-lab/clinvar/blob/master/clinvar_example_750_rows.tsv

bw2 replied:
After this, I would also get rid of the `variant_type` and `variant_type_change` columns as this info is in hgvs_c and hgvs_p

bw2 replied:
and for the same reason, get rid of Measure_type

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -20,43 +33,100 @@ def replace_semicolons(s, replace_with=":"):
-        'chrom', 'pos', 'ref', 'alt', 'mut', 'measureset_id', 'all_submitters', 'all_traits', 'all_pmids',
-        'inheritance_modes', 'age_of_onset', 'prevalence', 'disease_mechanism', 'xrefs'
+        'chrom', 'pos', 'ref', 'alt', 'mut', 'Measureset_type','Measureset_id','number_alleles','Measure_type','alleleid', 'name','gene_symbol',
+        'molecular_consequence','variant_type','variant_type_change','all_conditions_name'
```
would rename `all_condition_name` back to `all_traits`

XiaoleiZ commented on src/parse_clinvar_xml.py:
```diff
@@ -7,9 +11,10 @@
 from collections import defaultdict
 import xml.etree.ElementTree as ET
 
```
xml parse

mianasbat commented on src/parse_clinvar_xml.py:
```diff
@@ -128,21 +273,26 @@ def parse_clinvar_tree(handle,dest=sys.stdout,verbose=True,mode='collapsed'):
+    error.write("Done\n")
 
 
 def get_handle(path):
```
testing comment :)

XiaoleiZ commented on README.md:
```diff
@@ -2,66 +2,24 @@
-+ `conflicted` is `1` if the variant has *ever* been asserted "Pathogenic" or "Likely pathogenic" by any submitter for any phenotype, and has also been asserted "Benign" or "Likely benign" by any submitter for any phenotype, and `0` otherwise. Note that having one assertion of pathogenic and one of uncertain significance does *not* count as conflicted for this column. 
-
-To run the pipeline:
+#### Usage
```
test comment

XiaoleiZ committed:
Update parsing and joining

After parsing, there are two types of tables:
1. variant-condition specific
2. variant-specific

XiaoleiZ committed:
Update README.md

XiaoleiZ committed:
Update README.md

XiaoleiZ committed:
Update join_data.R

XiaoleiZ commented:
Changes made.

bw2 commented on src/clinvar_table_to_vcf.py:
```diff
@@ -90,12 +90,11 @@ def table_to_vcf(input_table_path):
-                    "inheritance_modes", "age_of_onset", "prevalence", 
-                    "disease_mechanism", "origin",
-                    "pathogenic", "conflicted", "benign"]:
+        for key in ['chrom','pos','ref','alt','mut','allele_id','measureset_type','measureset_id',
```
I would take out 'chrom', 'pos', 'ref', 'alt' from this list - so that they don't go into the INFO field.

bw2 commented on src/dedup_clinvar.py:
```diff
@@ -10,29 +10,31 @@
+#modified by Xioalei Zhang
 '''
-Run through a sorted clinvar_table.tsv file from the parse_clinvar_xml script, and make it unique on CHROM POS REF ALT
+Run through a sorted clinvar_table.tsv file from the parse_clinvar_xml script, and make it unique on CHROM POS REF ALT AlleleID
```
there's a 1-to-1 mapping from Allele ids to chrom-pos-ref-alt, so I would remove AlleleID from the uniqueness constraint. 

There are 3 records that are exceptions

15	91310153	ATCTGA	TAGATTC
17	7126557	G	A
X	25031777	GCC	GCCGCC

but they appear to be weird edge cases.

bw2 commented on src/master.py:
```diff
@@ -101,16 +102,28 @@ def download_if_changed(job_runner, local_path, ftp_host, ftp_path):
+
 # join information from the tab-delimited summary to the normalized genomic coordinates
-job.add("Rscript IN:join_data.R IN:%s" % variant_summary_table, input_filenames=['clinvar_table_normalized.tsv'], output_filenames=['clinvar_combined.tsv'])
+job.add("Rscript IN:join_data.R IN:%s" % variant_summary_table, input_filenames=['clinvar_record.tsv'], output_filenames=['clinvar_combined.tsv'])
```
I would remove the join_data.R step and delete that script now that all fields are being parsed out of the xml.

XiaoleiZ replied:
I think the variant_summary.txt file is still useful. Since it has the aggregated clinical_significance and review_status over all traits for the single allele, we could just make use of it by joining instead of computing from allele-trait table again.

bw2 replied:
That's a great point. Never mind then.

bw2 commented on README.md:
```diff
@@ -20,10 +20,10 @@ To create a flat representation of ClinVar suited for our purposes, we took seve
+3. Sort on genomic coordinates (we use GRCh37) 
+4. Normalize using [our Python implementation](https://github.com/ericminikel/minimal_representation/blob/master/normalize.py) of [vt normalize](http://genome.sph.umich.edu/wiki/Variant_Normalization) (see [[Tan 2015]]). The output file [clinvar_record.tsv.gz](clinvar_record.tsv.gz) after this step contains the variant-condition specific ClinVar records. Currently, the complex variations (e.g. a  ...
+5. Join the TXT file using [src/join_data.R](src/join_data.R) to aggregate interpretations from multiple submitters independent of condition. 
+6. Sort and de-duplicate  (this removes dups arising from duplicate records in the TXT dump). The output file [clinvar.tsv.gz](clinvar.tsv.gz) after this step contains the variant-specific aggregated data.
```
I like your idea of keeping a table with 1 record per allele-trait pair alongside the original table (which has 1 record per allele). 
Now that you're parsing everything from the XML, I think step 5 is no longer necessary - the several columns that it computes (like gold_stars) should just be computed in `parse_clinvar_xml.py`.

XiaoleiZ replied:
as join_data.R is still useful as talked above, I would keep it unchanged.

bw2 commented on README.md:
```diff
@@ -45,9 +45,10 @@ Other requirements are R, [htslib](https://github.com/samtools/htslib) and [vt](
-* [clinvar.tsv.gz](clinvar.tsv.gz)  
-* [clinvar.vcf.gz](clinvar.vcf)  
-* [clinvar_with_exac.tsv.gz](clinvar_with_exac.tsv.gz)  
+* [clinvar_record.tsv.gz](clinvar_record.tsv.gz): variant-condition specific record
```
to avoid people being confused about the difference between clinvar_record.tsv and clinvar.tsv, how about we call it `clinvar_allele_trait_pairs.tsv.gz` and the other file `clinvar_alleles.tsv.gz` instead of `clinvar.tsv.gz` (and then `clinvar_alleles_with_exac.tsv.gz`)

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -21,42 +22,89 @@ def remove_newlines_and_tabs(s):
+        current_row['allele_id']=''
+        
+        rcv = elem.find('./ReferenceClinVarAssertion/ClinVarAccession')
+        if(rcv.attrib.get('Type')!='RCV'):
```
style - wouldn't put parentheses around if condition

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -21,42 +22,89 @@ def remove_newlines_and_tabs(s):
                 if all(sequence_location.attrib.get(key) is not None for key in ('Chr', 'start', 'referenceAllele','alternateAllele')):
                     grch37_location = sequence_location
-
+                    break;
```
good idea to break. 
style - wouldn't put `;`

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -65,9 +113,44 @@ def parse_clinvar_tree(handle,dest=sys.stdout,verbose=True,mode='collapsed'):
+        genesymbol = elem.findall('.//Symbol')
+        if not genesymbol:
+            skipped_counter['missing gene symbol'] += 1
+            elem.clear()
```
why not keep records that don't have a gene symbol?

bw2 commented on src/parse_clinvar_xml.py:
```diff
@@ -96,11 +179,31 @@ def parse_clinvar_tree(handle,dest=sys.stdout,verbose=True,mode='collapsed'):
+            skipped_counter['missing clicinical significance description'] += 1
+            elem.clear()
+            continue 
+        if clinical_significance.find('.//ReviewStatus') is not None:
```
would save the results of .find() in a variable as previously

bw2 commented on src/dedup_clinvar.py:
```diff
@@ -10,29 +10,31 @@
 # recommended usage:
 # ./dedup_clinvar.py < clinvar_table_sorted.tsv > clinvar_table_dedup.tsv
```
This script is no longer really used for deduping. May be better to rename it to something like `group_by_allele.py`

bw2 commented on clinvar_example_750_rows.tsv:
```diff
@@ -1,750 +1,750 @@
-1	2160554	G	C	ALT	224868	SKI	Pathogenic	1	0	0	criteria provided, single submitter	1	NM_003036.3:c.349G>C	NP_003027.1:p.Gly117Arg	Baylor-Hopkins Center for Mendelian Genomics,Johns Hopkins University	Shprintzen-Goldberg syndrome	20301454,23023332		Infancy	<1 / 1 000 000		de novo	GeneReviews:NBK1277;Genetic Alliance:Shprintzen-Goldberg+craniosynostosis+syndrome/6581;Genetics Home Reference:shprintz ...
-1	2160557	G	A	ALT	222819	SKI	Likely pathogenic	1	0	0	criteria provided, single submitter	1	NM_003036.3:c.352G>A	NP_003027.1:p.Glu118Lys	Blueprint Genetics	Shprintzen-Goldberg syndrome	20301454		Infancy	<1 / 1 000 000		germline	GeneReviews:NBK1277;Genetic Alliance:Shprintzen-Goldberg+craniosynostosis+syndrome/6581;Genetics Home Reference:shprintzen-goldberg-syndrome;MedGen:C1321551;OMIM:182212;Off ...
-1	2160565	C	T	ALT	213682	SKI	Benign	0	1	0	criteria provided, single submitter	1	NM_003036.3:c.360C>T	NP_003027.1:p.Arg120=	GeneDx	not specified;not specified						germline	MedGen:CN169374
-1	2160622	C	T	ALT	213683	SKI	Benign	0	1	0	criteria provided, multiple submitters, no conflicts	2	NM_003036.3:c.417C>T	NP_003027.1:p.Ile139=	GeneDx;Invitae	not specified;Shprintzen-Goldberg syndrome	20301454		Infancy	<1 / 1 000 000		germline	OMIM:182212;GeneReviews:NBK1277;MedGen:C1321551;Office of Rare Diseases:4861;Genetic Alliance:Shprintzen-Goldberg+craniosynostosis+syndrome/6581;Orphanet:2462 ...
```
would change `molecular_consequence` 
from `NM_005101.3:c.163C>T:nonsense` => `nonsense`
since it's redundant with the hgvsc column

bw2 replied:
`origin` column is empty for all records in `clinvar_table_raw.tsv`. parsing may need to be fixed.

XiaoleiZ replied:
For molecular_consequence, it may have more than one values depending on the transcript. 
e.g.  [https://www.ncbi.nlm.nih.gov/clinvar/RCV000170331/](url)
Thus I prefer to keep this way of recording.

bw2 replied:
that makes sense.

bw2 commented on src/master.py:
```diff
@@ -101,16 +102,28 @@ def download_if_changed(job_runner, local_path, ftp_host, ftp_path):
 
-# now de-dup _again_, because the tab-delimited summary contains dups
-job.add("python -u IN:dedup_clinvar.py < IN:clinvar_combined_sorted.tsv | tee clinvar.tsv | bgzip -c > OUT:clinvar.tsv.gz")  # clinvar_combined_sorted_dedup.tsv.gz
+job.add("bgzip -c IN:clinvar.tsv > OUT:clinvar.tsv.gz")
```
in line #156 below (https://github.com/macarthur-lab/clinvar/pull/21/files#diff-ef363ae6652f439a7d0eafcf1c24ddeaR156), update column numbers to exclude measureset_id and add any new columns that have fewer than ~100 possible values.

XiaoleiZ commented on README.md:
```diff
@@ -20,10 +20,10 @@ To create a flat representation of ClinVar suited for our purposes, we took seve
+3. Sort on genomic coordinates (we use GRCh37) 
+4. Normalize using [our Python implementation](https://github.com/ericminikel/minimal_representation/blob/master/normalize.py) of [vt normalize](http://genome.sph.umich.edu/wiki/Variant_Normalization) (see [[Tan 2015]]). The output file [clinvar_record.tsv.gz](clinvar_record.tsv.gz) after this step contains the variant-condition specific ClinVar records. Currently, the complex variations (e.g. a  ...
+5. Join the TXT file using [src/join_data.R](src/join_data.R) to aggregate interpretations from multiple submitters independent of condition. 
+6. Sort and de-duplicate  (this removes dups arising from duplicate records in the TXT dump). The output file [clinvar.tsv.gz](clinvar.tsv.gz) after this step contains the variant-specific aggregated data.
```
As the reason above

XiaoleiZ committed:
Update

XiaoleiZ committed:
update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update

XiaoleiZ committed:
Update function name

XiaoleiZ committed:
Update the file formate
